### DISCUSSION TITLE
How to push selected commits?

### DISCUSSION THREAD

**Discussion by emptyfist** [community/community#30819] [Pull Requests] [answered]
I've one questions about Push in Git hub.

I made several commits(not pushed) on my side but some commits not necessary to push.
Atm, how can I push necessary few commits not all?

Always appreciate your help.

**Comment by airtower-luna** [comment, marked as answer]
If you want to push a commit you have to push all it's ancestors, too. That's part of how Git works, it's supposed to keep a persistent history.

If you want to [modify the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) before pushing it that's possible, though. After isn't *technically* impossible, but it gets harder (and possibly practically impossible) depending on where the commits end up.

**Reply by emptyfist** [reply]
Thank you for your reply.
So you it's impossible to push some necessary commits?

**Reply by airtower-luna** [reply]
It depends on how your commits are connected. Commits form a history. If your branch is at commit A and you make a new commit B, A is the *ancestor* of B. You cannot push a commit without it's ancestor, so you have to push the whole chain back to the initial commit.

If you have branches with different commits, you can decide which branch(es) to push. But if you decide to push a branch, you have to push its whole history.

See also: [Git Branching - Branches in a Nutshell](https://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell)
